Python/ceval.c

```diff
@@ -5020,6 +4649,130 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
+                    {
+                        // bpo-28598: Fast path for string formatting (but not
+                        // if the RHS is a str subclass).
+                        res = PyUnicode_Format(lhs, rhs);
+                        break;
+                    }
+                    res = PyNumber_Remainder(lhs, rhs);
+                    break;
+                case NB_OR:
+                    res = PyNumber_Or(lhs, rhs);
+                    break;
+                case NB_POWER:
+                    res = PyNumber_Power(lhs, rhs, Py_None);
+                    break;
+                case NB_RSHIFT:
+                    res = PyNumber_Rshift(lhs, rhs);
+                    break;
+                case NB_SUBTRACT:
+                    res = PyNumber_Subtract(lhs, rhs);
+                    break;
+                case NB_TRUE_DIVIDE:
+                    res = PyNumber_TrueDivide(lhs, rhs);
+                    break;
+                case NB_XOR:
+                    res = PyNumber_Xor(lhs, rhs);
+                    break;
+                case NB_INPLACE_ADD:
+                    res = PyNumber_InPlaceAdd(lhs, rhs);
+                    break;
+                case NB_INPLACE_AND:
+                    res = PyNumber_InPlaceAnd(lhs, rhs);
+                    break;
+                case NB_INPLACE_FLOOR_DIVIDE:
+                    res = PyNumber_InPlaceFloorDivide(lhs, rhs);
+                    break;
+                case NB_INPLACE_LSHIFT:
+                    res = PyNumber_InPlaceLshift(lhs, rhs);
+                    break;
+                case NB_INPLACE_MATRIX_MULTIPLY:
+                    res = PyNumber_InPlaceMatrixMultiply(lhs, rhs);
+                    break;
+                case NB_INPLACE_MULTIPLY:
+                    res = PyNumber_InPlaceMultiply(lhs, rhs);
+                    break;
+                case NB_INPLACE_REMAINDER:
+                    res = PyNumber_InPlaceRemainder(lhs, rhs);
+                    break;
+                case NB_INPLACE_OR:
+                    res = PyNumber_InPlaceOr(lhs, rhs);
+                    break;
+                case NB_INPLACE_POWER:
+                    res = PyNumber_InPlacePower(lhs, rhs, Py_None);
+                    break;
+                case NB_INPLACE_RSHIFT:
+                    res = PyNumber_InPlaceRshift(lhs, rhs);
+                    break;
+                case NB_INPLACE_SUBTRACT:
+                    res = PyNumber_InPlaceSubtract(lhs, rhs);
+                    break;
+                case NB_INPLACE_TRUE_DIVIDE:
+                    res = PyNumber_InPlaceTrueDivide(lhs, rhs);
+                    break;
+                case NB_INPLACE_XOR:
+                    res = PyNumber_InPlaceXor(lhs, rhs);
+                    break;
+                default:
+                    Py_UNREACHABLE();
+            }
+            if (res == NULL) {
+                goto error;
+            }
+            Py_DECREF(lhs);
+            Py_DECREF(rhs);
```

Comment:
You need to decref lhs and rhs before the error check, and I suspect that's where the refleak is from too.

The code in ``goto error`` only decrefs from TOS down, but you already popped lhs and rhs at line 4655, so in the error path, it won't decref lhs and rhs.

Reply:
Only `rhs` has been popped, so it is just the `Py_DECREF(rhs);` that needs to be moved before the test.

Reply:
Woops, you're right. Looks like I need a better pair of glasses :). That also explains why the refleak is `[1, 1, 1, 1, 1, 1]` and not `[2, 2, 2,...]`

Reply:
Sorry @brandtbucher I misled you a little, see Mark's comment at https://github.com/python/cpython/pull/29482#discussion_r746634254 . Only `rhs` needs a decref before the error check, `lhs` should still decref aftef.

Reply:
Yeah, I know, I just moved both `Py_DECREF`s and the `SET_TOP` above the error check. They're both correct (I just like keeping the `Py_DECREF`s together, since it can be a bit confusing otherwise).

Reply:
🤦 I didn't see the `SET_TOP` and was wondering how a double decref didn't happen.

Anyways, thanks for merging the opcodes! I'm now far less worried about running out of them due to specialization.